controllers/utils.go

```diff
@@ -0,0 +1,82 @@
+/*
+Copyright 2019-2021 Google LLC.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controllers
+
+import (
+	"crypto/sha256"
+	"fmt"
+	"strings"
+)
+
+// maxNEGNameLength is the max length for namespace, name and
+// port for neg name.  63 - 8 (suffix hash) - 3 (hyphen connector) = 52
+var maxNEGNameLength = 52
+
+// NEG returns backend neg name based on the service namespace, name
+// and target port. NEG naming convention:
+//
+//   {namespace}-{name}-{service port}-{hash}
+//
+// Output name is at most 63 characters.
+func generateNegName(namespace string, name string, portStr string) string {
```

Comment:
Erik suggested that we could import `k8s-io/ingress-gce` and call [`Namer`](https://github.com/kubernetes/ingress-gce/blob/4cb04408a6266b5ea00d9567c6165b9235392972/pkg/utils/namer/namer.go) directly. I tried this at first and got [this error](https://github.com/kubernetes/kubernetes/issues/90358), suggesting using `k8s-io/ingress-gce` as a dependency is discouraged. There is a workaround linked in the issue but it's neater to move the functionalities here.